### ISSUE TITLE
SM_CNN: Refactor

### ISSUE THREAD

**Issue by rosequ** [castorini/Castor#48]
- use torchtext 
- support batching
- simplify code to support backprop over word embedding, linguistic features etc. (#13)
I'll work on this issue.

@Impavidity also noted that we are using `conv1d` but in reality we should be using `conv2d` [here](https://github.com/castorini/Castor/blob/master/sm_cnn/model.py#L39) since the input is in 2D space. It'd be interesting to compare the scores with this implementation.

**Comment by lintool** [comment]
We should think about model management as we move forward... as we start improving existing models, should we create new directories or just add command-line options to old models?

Eventually sm_model would actually mean sm_model + {our few custom features}. I don't like `sm_model_v1`, `sm_model_v2`, but we need to be able to fish out the old models easily for baseline comparisons.

**Comment by rosequ** [comment]
I was thinking about this too. I think one of the solutions is to have is to support different modes. Something like `--mode vanilla` or `--mode linguistic`.

**Comment by lintool** [comment]
That would be an option, but then we need to keep some way to make sure new options don't break old code... and over time we'll accumulate cruft.

**Comment by rosequ** [comment]
Right, this is only if the network remains the same. If the structure of the network varies completely, we can different pairs of _network_ and _model_ folders in the same sm_cnn folder.

**Comment by snapbug** [comment]
I mean, I might be being an idiot and missing something, but doesn't the whole point of using version control solve this? When you want to refer to that version you can always refer to the commit id in the paper? If your experiment refers to multiple versions within the repo, then you can point to a script that will checkout the versions and run them appropriately. I mean, you should be doing this _anyway_.

Am I missing something?

**Comment by lintool** [comment]
Sure, but we need to remember commit id or cut "releases". In practice, though, what happens is that experiments get run over a couple weeks on different commit points, so issues slowly creep in.

Also, if we rely on commits, we have to checkout a particular commit point to run baselines, which will be kinda a pain...

**Comment by snapbug** [comment]
There's always tags that would give readable names without having to "release". The simplest solution is to run all the experiments once it's been determined what the experiments are. Either from the same commit, or writing a script that will checkout the versions that the experiment was run on.

I also don't really buy that checking out a commit is a pain, `git checkout <sha/tag>; <run experiment>; git checkout origin/master`. Also if the results are reproducible (as they should be) then you shouldn't _need_ to re-run baseline experiments unless the dataset has changed, and then you have to re-run them anyway.

**Comment by Impavidity** [comment]
As mentioned by @rosequ , the word embedding is actually tuned during the training in SM model (required_grad=False is missed in the code)
Furthermore, the logsoftmax is used in sm model in the final layer, but according to the paper, softmax is used.
But which implementation is better, it still depends on experiments.
